File: Mcs-cs/Mcs/Algorithms/McSplit.cs
using Mcs.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Mcs.Algorithms
{
    public class McSplit : MaximumCommonSubgraphAlgorithm
    {
        private Graph G1;
        private Graph G2;
        private List<Pair> BestM;

        public MaximumCommonSubgraph Compute(Graph g1, Graph g2)
        {
            G1 = g1;
            G2 = g2;
            BestM = new List<Pair>();
            Search(LabelClasses.From(g1, g2), new List<Pair>());
            return new MaximumCommonSubgraph(g1, g2, BestM);
        }

        private void Search(LabelClasses labelClasses, List<Pair> M)
        {
            if (M.Count > BestM.Count)
            {
                BestM = new List<Pair>(M);
            }

            int bound = M.Count + labelClasses.SumPairsOfVertices();
            if (bound <= BestM.Count)
            {
                return;
            }

            LabelClass selectedLabelClass = labelClasses.Select();
            int v = selectedLabelClass.SelectVertexFromG();
            foreach (int w in selectedLabelClass.H)
            {
                LabelClasses newLabelClasses = LabelClasses.Empty();
                for (int i = 0; i < labelClasses.Count; i++)
                {
                    List<int> neighborsOfV = labelClasses[i].NeighborsFromGOf(v);
                    List<int> neighborsOfW = labelClasses[i].NeighborsFromHOf(w);
                    if (neighborsOfV.Any() && neighborsOfW.Any())
                    {
                        newLabelClasses.Add(new LabelClass(G1, G2, neighborsOfV, neighborsOfW));
                    }
                    List<int> notNeighborsOfV = labelClasses[i].NotNeighborsFromGOf(v);
                    List<int> notNeighborsOfW = labelClasses[i].NotNeighborsFromHOf(w);
                    if (notNeighborsOfV.Any() && notNeighborsOfW.Any())
                    {
                        newLabelClasses.Add(new LabelClass(G1, G2, notNeighborsOfV, notNeighborsOfW));
                    }
                }
                M.Add(new Pair(v, w));
                Search(newLabelClasses, M);
                M.RemoveAt(M.Count - 1);
            }
            labelClasses.Remove(selectedLabelClass);
            if (selectedLabelClass.G.Count > 1)
            {
                labelClasses.Add(selectedLabelClass.CopyWithoutVertexFromG(v));
            }
            if (labelClasses.Count > 0)
            {
                Search(labelClasses, M);
            }
        }

        class LabelClass
        {
            private readonly Graph G1;
            private readonly Graph G2;

            public readonly List<int> G;
            public readonly List<int> H;
            public readonly int? MaxDegreeG;
            public readonly int? MaxDegreeVertexG;

            public LabelClass(Graph g1, Graph g2, List<int> G, List<int> H)
            {
                G1 = g1;
                G2 = g2;
                this.G = G;
                this.H = H;

                if (G.Count == 0)
                {
                    MaxDegreeG = null;
                    MaxDegreeVertexG = null;
                }
                else
                {
                    int maxDegreeG = g1.DegreeOf(G[0]);
                    int maxDegreeVertexG = G[0];
                    for (int i = 1; i < G.Count; i++)
                    {
                        int degree = g1.DegreeOf(G[i]);
                        if (degree > maxDegreeG)
                        {
                            maxDegreeG = degree;
                            maxDegreeVertexG = G[i];
                        }
                    }
                    MaxDegreeG = maxDegreeG;
                    MaxDegreeVertexG = maxDegreeVertexG;
                }
            }

            public LabelClass CopyWithoutVertexFromG(int vertex)
            {
                return new LabelClass(G1, G2, G.Where(v => v != vertex).ToList(), H);
            }

            public int SelectVertexFromG()
            {
                if (MaxDegreeVertexG == null)
                {
                    throw new SystemException("No vertices to select from in G!");
                }
                return (int)MaxDegreeVertexG;
            }

            public List<int> NeighborsFromGOf(int vertex)
            {
                return G.Where(v => G1.AdjacencyMatrix[vertex][v] == 1).ToList();
            }

            public List<int> NeighborsFromHOf(int vertex)
            {
                return H.Where(v => G2.AdjacencyMatrix[vertex][v] == 1).ToList();
            }

            public List<int> NotNeighborsFromGOf(int vertex)
            {
                return G.Where(v => v != vertex && G1.AdjacencyMatrix[vertex][v] == 0).ToList();
            }

            public List<int> NotNeighborsFromHOf(int vertex)
            {
                return H.Where(v => v != vertex && G2.AdjacencyMatrix[vertex][v] == 0).ToList();
            }

            public override string ToString()
            {
                return base.ToString();
            }
        }

        class LabelClasses
        {
            private readonly List<LabelClass> Items;

            public LabelClasses(List<LabelClass> labelClasses)
            {
                Items = labelClasses;
            }

            public static LabelClasses Empty()
            {
                return new LabelClasses(new List<LabelClass>());
            }

            public static LabelClasses From(Graph g1, Graph g2)
            {
                List<int> G = Enumerable.Range(0, g1.N).ToList();
                List<int> H = Enumerable.Range(0, g2.N).ToList();

                return new LabelClasses(new List<LabelClass>() { new LabelClass(g1, g2, G, H) });
            }

            public LabelClass this[int index]
            {
                get
                {
                    return Items[index];
                }
            }

            public int Count
            {
                get
                {
                    return Items.Count;
                }
            }

            public void Add(LabelClass labelClass)
            {
                Items.Add(labelClass);
            }

            public void Remove(LabelClass labelClass)
            {
                if (!Items.Remove(labelClass))
                {
                    throw new SystemException("Unable to remove label class!");
                }
            }

            public int SumPairsOfVertices()
            {
                return Items.Select(labelClass => Math.Min(labelClass.G.Count, labelClass.H.Count)).Sum();
            }

            /// <summary>
            ///  Selects a label class with the smallest max(|G|,|H|) and breaks ties by
            ///  selecting a class containing a vertex in G with the largest degree
            /// </summary>
            public LabelClass Select()
            {
                LabelClass best = Items[0];
                int maxVerticesBest = Math.Max(best.G.Count, best.H.Count);
                foreach (LabelClass labelClass in Items)
                {
                    int maxVerticesLabelClass = Math.Max(labelClass.G.Count, labelClass.H.Count);
                    if (maxVerticesBest < maxVerticesLabelClass)
                    {
                        best = labelClass;
                        maxVerticesBest = maxVerticesLabelClass;
                    }
                    else if (maxVerticesLabelClass == maxVerticesBest)
                    {
                        if (best.MaxDegreeG < labelClass.MaxDegreeG)
                        {
                            best = labelClass;
                            maxVerticesBest = maxVerticesLabelClass;
                        }
                    }
                }
                return best;
            }

            public override string ToString()
            {
                return base.ToString();
            }
        }

    }
}

File: mcs-java/src/main/java/anc/algorithms/McSplit.java
package anc.algorithms;

import static com.google.common.collect.Lists.newArrayList;
import static java.lang.String.format;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import anc.models.Graph;
import anc.models.MaximumCommonSubgraph;
import anc.models.Pair;

public class McSplit implements MaximumCommonSubgraphAlgorithm {
    private Graph g1;
    private Graph g2;
    private List<Pair> bestM;

    @Override
    public MaximumCommonSubgraph compute(Graph g1, Graph g2) {
        this.g1 = g1;
        this.g2 = g2;
        bestM = newArrayList();
        search(LabelClasses.from(g1, g2), newArrayList());
        final List<Pair> sortedM = bestM.stream().sorted(Comparator.comparingInt(p -> p.v1)).collect(Collectors.toList());
        return new MaximumCommonSubgraph(sortedM, g1, g2);
    }

    private void search(LabelClasses labelClasses, List<Pair> M) {
        if (M.size() > bestM.size()) {
            bestM = newArrayList(M);
        }

        int bound = M.size() + labelClasses.sumPairsOfVertices();
        if (bound <= bestM.size()) {
            return;
        }

        LabelClass selectedLabelClass = labelClasses.select();
        int v = selectedLabelClass.selectVertexFromG();
        for (int w : selectedLabelClass.H) {
            LabelClasses newLabelClasses = LabelClasses.empty();
            for (LabelClass labelClass : labelClasses.labelClasses) {
                List<Integer> neighborsOfV = labelClass.neighborsFromGOf(v);
                List<Integer> neighborsOfW = labelClass.neighborsFromHOf(w);
                if (!neighborsOfV.isEmpty() && !neighborsOfW.isEmpty()) {
                    newLabelClasses.add(new LabelClass(g1, g2, neighborsOfV, neighborsOfW));
                }
                List<Integer> notNeighborsOfV = labelClass.notNeighborsFromGOf(v);
                List<Integer> notNeighborsOfW = labelClass.notNeighborsFromHOf(w);
                if (!notNeighborsOfV.isEmpty() && !notNeighborsOfW.isEmpty()) {
                    newLabelClasses.add(new LabelClass(g1, g2, notNeighborsOfV, notNeighborsOfW));
                }
            }
            M.add(new Pair(v, w));
            search(newLabelClasses, M);
            M.remove(M.size() - 1);
        }
        labelClasses.remove(selectedLabelClass);
        if (selectedLabelClass.G.size() > 1) {
            labelClasses.add(selectedLabelClass.copyWithoutVertexFromG(v));
        }
        if (!labelClasses.labelClasses.isEmpty()) {
            search(labelClasses, M);
        }

    }

    static class LabelClass {
        private final Graph g1;
        private final Graph g2;

        public final List<Integer> G;
        public final List<Integer> H;
        public final Integer maxDegreeG;
        public final Integer maxDegreeVertexG;

        public LabelClass(Graph g1, Graph g2, List<Integer> G, List<Integer> H) {
            this.g1 = g1;
            this.g2 = g2;
            this.G = G;
            this.H = H;

            if (G.isEmpty()) {
                this.maxDegreeG = null;
                this.maxDegreeVertexG = null;
            } else {
                int maxDegreeG = g1.degreeOf(G.get(0));
                int maxDegreeVertexG = G.get(0);
                for (int i = 1; i < G.size(); i++) {
                    int degree = g1.degreeOf(G.get(i));
                    if (degree > maxDegreeG) {
                        maxDegreeG = degree;
                        maxDegreeVertexG = G.get(i);
                    }
                }
                this.maxDegreeG = maxDegreeG;
                this.maxDegreeVertexG = maxDegreeVertexG;
            }
        }

        public LabelClass copyWithoutVertexFromG(int vertex) {
            return new LabelClass(g1, g2, G.stream().filter(v -> v != vertex).collect(Collectors.toList()), H);
        }

        public int selectVertexFromG() {
            return maxDegreeVertexG;
        }

        public List<Integer> neighborsFromGOf(int vertex) {
            return G.stream().filter(v -> g1.M[vertex][v] == 1).collect(Collectors.toList());
        }

        public List<Integer> neighborsFromHOf(int vertex) {
            return H.stream().filter(v -> g2.M[vertex][v] == 1).collect(Collectors.toList());
        }

        public List<Integer> notNeighborsFromGOf(int vertex) {
            return G.stream().filter(v -> v != vertex && g1.M[vertex][v] == 0).collect(Collectors.toList());
        }

        public List<Integer> notNeighborsFromHOf(int vertex) {
            return H.stream().filter(v -> v != vertex && g2.M[vertex][v] == 0).collect(Collectors.toList());
        }

        @Override
        public String toString() {
            return "LabelClass [G=" + G + ", H=" + H + "]";
        }
    }

    static class LabelClasses {
        public final List<LabelClass> labelClasses;

        public LabelClasses(List<LabelClass> labelClasses) {
            this.labelClasses = labelClasses;
        }

        public static LabelClasses empty() {
            return new LabelClasses(new ArrayList<>());
        }

        public static LabelClasses from(Graph g1, Graph g2) {
            List<Integer> G = IntStream.range(0, g1.n).boxed().collect(Collectors.toList());
            List<Integer> H = IntStream.range(0, g2.n).boxed().collect(Collectors.toList());
            return new LabelClasses(newArrayList(new LabelClass(g1, g2, G, H)));
        }

        public void add(LabelClass labelClass) {
            labelClasses.add(labelClass);
        }

        public void remove(LabelClass labelClass) {
            if (!labelClasses.remove(labelClass)) {
                throw new RuntimeException(format("Unable to remove label class %s!", labelClass));
            }
        }

        public int sumPairsOfVertices() {
            int sum = 0;
            for (LabelClass labelClass : labelClasses) {
                sum += Math.min(labelClass.G.size(), labelClass.H.size());
            }
            return sum;
        }

        /**
         * Selects a label class with the smallest max(|G|,|H|) and breaks ties by
         * selecting a class containing a vertex in G with the largest degree
         */
        public LabelClass select() {
            // TODO: Speed up
            LabelClass best = labelClasses.get(0);
            int maxVerticesBest = Math.max(best.G.size(), best.H.size());
            for (LabelClass labelClass : labelClasses) {
                int maxVerticesLabelClass = Math.max(labelClass.G.size(), labelClass.H.size());
                if (maxVerticesBest < maxVerticesLabelClass) {
                    best = labelClass;
                    maxVerticesBest = maxVerticesLabelClass;
                } else if (maxVerticesLabelClass == maxVerticesBest) {
                    if (best.maxDegreeG < labelClass.maxDegreeG) {
                        best = labelClass;
                        maxVerticesBest = maxVerticesLabelClass;
                    }
                }
            }
            return best;
        }

        @Override
        public String toString() {
            return "LabelClasses [labelClasses=" + labelClasses + "]";
        }
    }
}
